######################################
#  R Code used for data analysis and simulations in
#  G.Margolin, A.Tang, and S.Leikin. Differential expression analysis in single cell and spatial RNASeq without model assumptions.
#  bioRxiv preprint: https://www.biorxiv.org/content/10.1101/2025.10.20.683496v1
######################################

####################################################
# Requirements for successful code execution:
# R-PACKAGES AND LIBRARIES: dplyr, tidyverse, and Seurat 5.0 or later (may work but not tested with earlier Seurat versions)
# DATA: All data must be assembled into a single Seurat object with a single counts layer within active "RNA" assay
#       The active assay can be simply renamed to "RNA", but we recommend assembling the Seurat object from count matrices
#       with Seurat's CreateSeuratObject() function. 
#       It is important to use Seurat's JoinLayers() function to ensure a single counts layer.
#
# FUNCTIONS: Necessary functions must be loaded by executing the entire sc&spRNASeqFunctions_final.R file,
# which can be downloaded from https://github.com/sergeyleikin/sc-sp-RNASeq
####################################################

####################################################
# DATASETS:
#'* Neuron.QC is a dataset for GABAergic and glutamatergic neurons assembled from 10X Genomics data, which is available for download at:*
# https://github.com/sergeyleikin/sc-sp-RNASeq
#
# 10k Mouse E18 Combined Cortex, Hippocampus and Subventricular Zone Cells, Single Indexed
# Universal 3' Gene Expression dataset analyzed using Cell Ranger 4.0.0, 10X Genomics, (2020, July 7)
# https://www.10xgenomics.com/datasets/10-k-mouse-e-18-combined-cortex-hippocampus-and-subventricular-zone-cells-single-indexed-3-1-standard-4-0-0
# 10k Mouse E18 Combined Cortex, Hippocampus and Subventricular Zone Cells, Dual Indexed
# Universal 3' Gene Expression dataset analyzed using Cell Ranger 4.0.0, 10X Genomics, (2020, July 7)
# https://www.10xgenomics.com/datasets/10-k-mouse-e-18-combined-cortex-hippocampus-and-subventricular-zone-cells-dual-indexed-3-1-standard-4-0-0.
#
# Neuron.QC was assembled as follows: Step 1. The original filtered HDF5 files were read into Seurat followed by
# quality control (QC) based on 0.75 < % mitochondrial mRNA < 7.5 and total UMI count > 1000. Step 2. 
# The QC thresholds were selected based on the appearance of the corresponding violin plots. The resulting Seurat objects were merged.
# Step 3. Two objects containing just GABAergic and just glutamatergic neuron were created by using the subset() function based on expression of
# the corresponding marker genes described in the paper, assigned identities "GABA" and "GLUT", and merged into a single Seurat object followed by
# joining its layers using the JoinLayers() function.
#
#'* E18tib.PCN is a preassembled spRNASeq dataset for proximal proliferative chondrocyte areas in 3 different sections of the same E18tibia available for download at:*
# https://github.com/sergeyleikin/sc-sp-RNASeq
# The full E18 tibia dataset is available for download at NCBI GEO (https://www.ncbi.nlm.nih.gov/geo), accession number GSE299816
# Sample Seurat-compatible functions for reading and cropping VisiumHD datasets and *.csv files for the 3 proliferating chondrocyte area selections
# in proximal tibia generated by Loupe Browser (10X Genomics) are available for download at: https://github.com/sergeyleikin/sc-sp-RNASeq
# Quality control parameters for generating E18tib.PCN were: > 20 features per bin, > 100 counts other than Col1a1 or Col1a2 per bin, mitochondrial genes
# between 0.1 and 5% UMI, and Col2a1 > 5% UMI. Seurat's relative count normalization was utilized to generate the data layer of E18tib.PCN
#
#####################################################
#     READ AND NORMALIZE Neuron.QC
#####################################################
#
Neuron.QC<-readRDS("Neuron.QC.rds") 
Neuron.QCN<-NormalizeData(Neuron.QC, normalization.method = "RC")  #relative count normalization
Neuron.QCLN<-NormalizeData(Neuron.QC)                              #lognormalization
#
######################################################
#'* FIGURE 1*
######################################################
#
# Fig. 1A
VlnPlot(Neuron.QC, features = "nCount_RNA", log = FALSE)
#
# Fig. 1B
GABA.QC=subset(Neuron.QC, idents = "GABA")
Ni.GABA=colSums(GABA.QC[["RNA"]]$counts)
Nc.GABA=ncol(GABA.QC[["RNA"]]$counts)
w.AC.GABA<-Ni.GABA/sum(Ni.GABA)
Var.GABA=var(w.AC.GABA)
Ci.GABA<-as.matrix(GABA.QC[["RNA"]]$counts)
wzi.GABA<-Ci.GABA                             #initialize wzi matrix
WDCplot<-matrix(ncol=2,nrow=nrow(Ci.GABA))
rownames(WDCplot)=rownames(Ci.GABA)
colnames(WDCplot)=c("Counts/Cell","WDC")
for (gene in rownames(Ci.GABA)) {
  wzi.GABA[gene,]=ICCWeight(h=Ci.GABA[gene,],n=Ni.GABA)
  WDCplot[gene,1]=sum(Ci.GABA[gene,])/Nc.GABA
  WDCplot[gene,2]=var(wzi.GABA[gene,])/Var.GABA
}
write.csv(as.data.frame(WDCplot),"WDC.GABA.csv")
GLUT.QC=subset(Neuron.QC, idents = "GLUT")
Ni.GLUT=colSums(GLUT.QC[["RNA"]]$counts)
Nc.GLUT=ncol(GLUT.QC[["RNA"]]$counts)
w.AC.GLUT<-Ni.GLUT/sum(Ni.GLUT)
Var.GLUT=var(w.AC.GLUT)
Ci.GLUT<-as.matrix(GLUT.QC[["RNA"]]$counts)
wzi.GLUT<-Ci.GLUT                             #initialize wzi matrix
WDCplot<-matrix(ncol=2,nrow=nrow(Ci.GLUT))
rownames(WDCplot)=rownames(Ci.GLUT)
colnames(WDCplot)=c("Counts/Cell","WDC")
for (gene in rownames(Ci.GLUT)) {
  wzi.GLUT[gene,]=ICCWeight(h=Ci.GLUT[gene,],n=Ni.GLUT)
  WDCplot[gene,1]=sum(Ci.GLUT[gene,])/Nc.GLUT
  WDCplot[gene,2]=var(wzi.GLUT[gene,])/Var.GLUT
}
write.csv(as.data.frame(WDCplot),"WDC.GLUT.csv")
#
######################################################
#'* FIGURE 2*
######################################################
#
# Gene list: 12,087 genes with at least 50 total counts in glutamatergic neurons 
#
GLUT.full.counts<-as.matrix(GLUT.QC[["RNA"]]$counts)
GLUT.counts<-GLUT.full.counts[rowSums(GLUT.full.counts)>=50,] #removing ~9,000 genes with fewer than 50 aggregate counts (<0.05 counts/cell)
GeneList=rownames(GLUT.counts)
#
# Fig. 2B
#
j<-0
gaba.glut.means<-data.frame(row.names = GeneList,gaba.mean=numeric(length(GeneList)), gaba.wmean=numeric(length(GeneList)),
                            glut.mean=numeric(length(GeneList)), glut.wmean=numeric(length(GeneList)))
for (gene in GeneList){
  j=j+1
  glut.h=Ci.GLUT[gene,]
  gaba.h=Ci.GABA[gene,]
  glut.x=100*glut.h/Ni.GLUT
  gaba.x=100*gaba.h/Ni.GABA
  glut.w=ICCWeight(h=glut.h,n=Ni.GLUT)
  gaba.w=ICCWeight(h=gaba.h,n=Ni.GABA)
  gaba.glut.means[gene,"glut.wmean"]<-sum(glut.x*glut.w)
  gaba.glut.means[gene,"glut.mean"]<-mean(glut.x)
  gaba.glut.means[gene,"gaba.wmean"]<-sum(gaba.x*gaba.w)
  gaba.glut.means[gene,"gaba.mean"]<-mean(gaba.x)
  print(paste("Gene number:",j))
}
rownames(gaba.glut.means)<-GeneList
write.csv(gaba.glut.means,"gaba.glut.means.csv")
#
# Fig. 2C
#
GLUT.GABA.T<-IterWghtTtest(Neuron.QC, features = GeneList, ident.1 = "GLUT",ident.2 = "GABA", min.count=0, max.pval = 1, icc=1) #unweighted t-test
GLUT.GABA.IWT<-IterWghtTtest(Neuron.QC, features=GeneList, ident.1 = "GLUT",ident.2 = "GABA", min.count=0, max.pval = 1, icc = "i") #weighted t-test
GLUT.GABA.WLK<-as.data.frame(FindMarkers(Neuron.QCN,features = GeneList, ident.1 = "GLUT",ident.2 = "GABA",logfc.threshold = 0,min.pct=0)) #Seurat's default
GLUT.GABA.T.05<-GLUT.GABA.T[GLUT.GABA.T$p.value<0.05,]
GLUT.GABA.WLK.05<-GLUT.GABA.WLK[GLUT.GABA.WLK$p_val<0.05,]
GLUT.GABA.IWT.05<-GLUT.GABA.IWT[GLUT.GABA.IWT$p.value<0.05,]
GLUT.GABA.T.0001<-GLUT.GABA.T.05[GLUT.GABA.T.05$p.value<0.0001,]
GLUT.GABA.IWT.0001<-GLUT.GABA.IWT[GLUT.GABA.IWT$p.value<0.0001,]
GLUT.GABA.WLK.0001<-GLUT.GABA.WLK.05[GLUT.GABA.WLK.05$p_val<0.0001,]
GLUT.GABA.WLK<-GLUT.GABA.WLK[GeneList,]
write.csv(GLUT.GABA.T,"GLUT.GABA.T.csv")
write.csv(GLUT.GABA.IWT,"GLUT.GABA.IWT.csv")
write.csv(GLUT.GABA.WLK,"GLUT.GABA.WLK.csv")
#Number of genes found by 3 tests with p<0.0001
nrow(GLUT.GABA.T.0001) #t-test -- 3,747
nrow(GLUT.GABA.IWT.0001) #weighted t-test -- 3,993
nrow(GLUT.GABA.WLK.0001) #Wilcoxon test -- 8,045
#
# Test comparisons
length(setdiff(rownames(GLUT.GABA.IWT.0001),rownames(GLUT.GABA.T.05))) #6 genes found by weighted-t test with p<0.0001 but not t-test with p<0.05
length(setdiff(rownames(GLUT.GABA.IWT.0001),rownames(GLUT.GABA.WLK.05))) #845 genes found by weighted-t with p<0.0001 but not Wilcoxon test with p<0.05
length(setdiff(rownames(GLUT.GABA.T.0001),rownames(GLUT.GABA.IWT.05))) #3 genes found by t-test with p<0.0001 but not weighted-t with p<0.05
length(setdiff(rownames(GLUT.GABA.WLK.0001),rownames(GLUT.GABA.IWT.05))) #3,523 genes found by Wilcoxon test with p<0.0001 but not weighted-t with p<0.05
#
# Fig. 2A
#
log2Fcompare<-matrix(nrow = length(GeneList), ncol = 2)
rownames(log2Fcompare)<-GeneList
colnames(log2Fcompare)<-c("Common","weighted")
log2Fcompare[,1]<-GLUT.GABA.T$log2FC
log2Fcompare[,2]<-GLUT.GABA.IWT$log2FC
log2Fsigns<-log2Fcompare[log2Fcompare[,1]*log2Fcompare[,2]<0,]
nrow(log2Fsigns)*100/12087
log2F025<-log2Fsigns[abs(log2Fsigns[,1]-log2Fsigns[,2])>0.25,]
nrow(log2F025)*100/847
log2F05<-log2Fsigns[abs(log2Fsigns[,1]-log2Fsigns[,2])>0.5,]
nrow(log2F05)*100/847
#
######################################################
#'* FIGURE 3A-G*
######################################################
#
# Fig. 3F: Effect of SCTransform on glutamatergic neurons 
#
#Performing SCTransform on unaltered neuron data
#
Neuron.SCT<-SCTransform(Neuron.QC,vars.to.regress = "percent.mt", verbose = FALSE)
#
#Reassembling GLUT.SCT object from SCTransform renormalized counts, glutamatergic neurons only
#and GLUT.RNA object from original counts with the same set of genes as GLUT.SCT
#
GLUT.SCT<-subset(Neuron.SCT, idents = "GLUT")
GLUT.SCT.mat<-as.matrix(GLUT.SCT[["SCT"]]$counts)
genes.SCT<-rownames(GLUT.SCT.mat)
GLUT.RNA.mat<-as.matrix(GLUT.SCT[["RNA"]]$counts)
GLUT.RNA.mat<-GLUT.RNA.mat[genes.SCT,]
GLUT.SCT<-CreateSeuratObject(counts = GLUT.SCT.mat)
GLUT.RNA<-CreateSeuratObject(counts = GLUT.RNA.mat)
#
#Analyzing effects of SCTransform
#
Idents(GLUT.SCT)<-"GLUT.SCT"
Idents(GLUT.RNA)<-"GLUT.RNA"
GLUT.SCTM<-JoinLayers(merge(GLUT.SCT, y=c(GLUT.RNA), add.cell.ids=c("C1","C2")))
GLUT.SCT.IWT<-IterWghtTtest(GLUT.SCTM, ident.1 = "GLUT.SCT", ident.2 = "GLUT.RNA")
write.csv(GLUT.SCT.IWT,"GLUT.SCT.IWT.csv")
GLUT.SCT.IWT.0001<-GLUT.SCT.IWT[GLUT.SCT.IWT$p.value<0.001,]
GLUT.SCT.IWT.00001<-GLUT.SCT.IWT[GLUT.SCT.IWT$p.value<0.0001,]
nrow(GLUT.SCT.IWT.0001) # 854 false DGE with p<0.0001 created by SCTransform
nrow(GLUT.SCT.IWT.00001) # 720 false DGE with p<0.00001 created by SCTransform
#
# Fig. 3A-E & Fig. 3G: Effects of random count sub-sampling 
#
Ng<-nrow(GLUT.full.counts)
Nc<-ncol(GLUT.full.counts)
#Random count subsampling
subs.F=0.4
GLUTs.Allcnt<-GLUT.full.counts # initializing matrix for count subsampling
#Subsampling counts
results.SS<-data.frame(FM001=numeric(10), FMT001=numeric(10), T001=numeric(10), CHIIWT001=numeric(10),
                       SCT001=numeric(10), FMSCT001=numeric(10))
for (sim in c(1:10)){
  for (j in c(1:Ng)) {
    set.seed((sim-1)*Ng+j)
    GLUTs.Allcnt[j,]=rbinom(Nc,GLUT.full.counts[j,],subs.F)  #count subsampling loop
  }
  print(paste("Simulation:", sim))
  GLUTm.Allcnt<-GLUT.full.counts[rowSums(GLUT.full.counts)>0,] #eliminating genes with 0 aggregate counts in original sample
  GLUTsm.Allcnt<-GLUTs.Allcnt[rowSums(GLUTs.Allcnt)>0,] #eliminating genes with 0 aggregate counts in subsampled sample  
  GLUTs.QC<-CreateSeuratObject(counts = GLUTsm.Allcnt, min.cells = 0, min.features = 0) #Creating Seurat objects
  GLUTo.QC<-CreateSeuratObject(counts = GLUTm.Allcnt, min.cells = 0, min.features = 0)
  GLUTs.QC[["percent.mt"]] <- PercentageFeatureSet(GLUTs.QC, pattern = "^mt-")
  GLUTo.QC[["percent.mt"]] <- PercentageFeatureSet(GLUTo.QC, pattern = "^mt-")
  GLUTs.SCT<-SCTransform(GLUTs.QC,vars.to.regress = "percent.mt", verbose = FALSE)
  GLUTo.SCT<-SCTransform(GLUTo.QC,vars.to.regress = "percent.mt", verbose = FALSE)
  GLUTs.SCT.mat<-as.matrix(GLUTs.SCT[["SCT"]]$counts)
  GLUTo.SCT.mat<-as.matrix(GLUTo.SCT[["SCT"]]$counts)
  GLUTs.SCT<-CreateSeuratObject(counts = GLUTs.SCT.mat, min.cells = 0, min.features = 0)
  GLUTo.SCT<-CreateSeuratObject(counts = GLUTo.SCT.mat, min.cells = 0, min.features = 0)
  Idents(GLUTs.QC)<-"Subset"
  Idents(GLUTo.QC)<-"Orig"
  Idents(GLUTs.SCT)<-"Subset.SCT"
  Idents(GLUTo.SCT)<-"Orig.SCT"
  GLUTso.QC<-JoinLayers(merge(GLUTo.QC, y=c(GLUTs.QC,GLUTs.SCT,GLUTo.SCT), add.cell.ids = c("C1","C2","C3","C4"), merge.data = "FALSE")) #merging Seurat objects 
  GLUTso.QCN<-NormalizeData(GLUTso.QC) #log-normalization
  GLUT30.counts<-GLUT.full.counts[rowSums(GLUT.full.counts)>30,] #identifying genes with >30 counts in original sample
  genes.s<-rownames(GLUT30.counts)
  GLUTso.FM <- as.data.frame(FindMarkers(GLUTso.QCN, features=genes.s ,ident.1 = "Orig",ident.2 = "Subset",logfc.threshold = 0,min.pct=0)) #Wilcoxon test
  GLUTso.FM.SCT <- as.data.frame(FindMarkers(GLUTso.QCN, features=genes.s ,ident.1 = "Orig.SCT",ident.2 = "Subset.SCT",logfc.threshold = 0,min.pct=0)) #SCT+Wilcoxon test
  GLUTso.DGE <- DGE.2samples(GLUTso.QC, features = genes.s, ident.1 = "Orig",ident.2 = "Subset", min.count=0, max.pval = 1) #weighted-t/chi^2 test
  GLUTso.T <- IterWghtTtest(GLUTso.QC, features = genes.s, ident.1 = "Orig",ident.2 = "Subset", min.count=0, max.pval = 1, icc = 1) #common t-test
  print("Performing FindMarkers t-test")   #FindMarkers t-test for log-normalized data
  GLUTso.FMT <- as.data.frame(FindMarkers(GLUTso.QCN, features=genes.s, test.use="t", ident.1 = "Orig",ident.2 = "Subset",logfc.threshold = 0,min.pct=0)) 
  GLUTso.DGE.SCT <- DGE.2samples(GLUTso.QC, features = genes.s, ident.1 = "Orig.SCT",ident.2 = "Subset.SCT", min.count=0, max.pval = 1) #SCT+weighted-t/chi^2-test
  #Simulation results
  GLUTso.FM.001<-GLUTso.FM[GLUTso.FM$p_val<0.001,]
  GLUTso.FM.SCT.001<-GLUTso.FM.SCT[GLUTso.FM.SCT$p_val<0.001,]
  GLUTso.DGE.001<-GLUTso.DGE[((GLUTso.DGE$p.value<0.001)&(GLUTso.DGE$Chi2.p.value<0.05)),]
  GLUTso.T.001<-GLUTso.T[GLUTso.T$p.value<0.001,]
  GLUTso.FMT.001<-GLUTso.FMT[GLUTso.FMT$p_val<0.001,]
  GLUTso.DGE.SCT.001<-GLUTso.DGE.SCT[((GLUTso.DGE.SCT$p.value<0.001)&(GLUTso.DGE.SCT$Chi2.p.value<0.05)),]
  results.SS$FM001[sim]=nrow(GLUTso.FM.001)
  results.SS$FMT001[sim]=nrow(GLUTso.FMT.001)
  results.SS$T001[sim]=nrow(GLUTso.T.001)
  results.SS$CHIIWT001[sim]=nrow(GLUTso.DGE.001)
  results.SS$SCT001[sim]=nrow(GLUTso.DGE.SCT.001)
  results.SS$FMSCT001[sim]=nrow(GLUTso.FM.SCT.001)
}
#Saving results as csv files
write.csv(GLUTso.FM,"GLUT.SS.FM.40%.csv")
write.csv(GLUTso.DGE,"GLUT.SS.DGE.40%.csv")
write.csv(GLUTso.T,"GLUT.SS.T.40%.csv")
write.csv(GLUTso.FMT,"GLUT.SS.FMT.40%.csv")
write.csv(GLUTso.DGE.SCT,"GLUT.SS.2T.SCT.40%.csv")
write.csv(GLUTso.FM.SCT,"GLUT.SS.FM.SCT.40%.csv")
write.csv(results.SS,"results.SS.40%.csv")
#
######################################################
#'* FIGURE 3H-N AND FIGURE 4*
######################################################
#
GLUT.counts=GLUT.full.counts
#
# For Fig. 3H-N use sim="Fig.3", for Fig. 4 use sim="Fig.4"
sim="Fig.4"
if(sim=="Fig.3") {Genes<-rownames(GLUT.counts[rowSums(GLUT.counts)>30,])}
if(sim=="Fig.4") {Genes<-rownames(GLUT.counts[(rowSums(GLUT.counts)/ncol(GLUT.counts)>0.3)&(rowSums(GLUT.counts)/ncol(GLUT.counts)<3),])}
#
results.RN<-data.frame(FM05=numeric(10),FM001=numeric(10),FMT05=numeric(10),FMT001=numeric(10),
                       T05=numeric(10),T001=numeric(10),CHIIWT05=numeric(10),CHIIWT001=numeric(10),
                       SCT05=numeric(10),SCT001=numeric(10))
results.BN<-data.frame(FM05s=numeric(10),FM05.true=numeric(10),FM001=numeric(10),FM001.true=numeric(10),FMT05s=numeric(10),FMT05.true=numeric(10),FMT001=numeric(10),FMT001.true=numeric(10),
                       T05s=numeric(10),T05.true=numeric(10),T001=numeric(10),T001.true=numeric(10),CHIIWT05s=numeric(10),CHIIWT05.true=numeric(10),
                       CHIIWT001=numeric(10),CHIIWT001.true=numeric(10),SCT05s=numeric(10),SCT05.true=numeric(10),SCT001=numeric(10),SCT001.true=numeric(10))
for (i in c(1:10)){
  set.seed(i)
  modexGenes<-sample(Genes,1000) #genes to be modified by noise
  GLUT.DFb<-GLUT.counts[modexGenes,] #count matrix for genes to be modified by noise
  Nr<-nrow(GLUT.DFb)
  modexGenes.mod<-sample(modexGenes,300) #genes to be modified by noise and change in expression
  DF<-sqrt(15) #noise scaling factor
  GLUT.DFb.bias<-0*GLUT.DFb #initialization of change in expression matrix
  #Bias definition (50% of counts)
  GLUT.DFb.bias[modexGenes.mod,]<-round(0.5*GLUT.DFb[modexGenes.mod,])
  #Adding noise
  set.seed(2*i)
  Unoise<-matrix(runif(Nc*Nr,min=-1,max=1),ncol=Nc,nrow=Nr) #uniform noise
  Tnoise<-round(Unoise*(DF-(DF-1)/(1+0.1*(GLUT.DFb-1)))*sqrt(GLUT.DFb)) #noise scaling function
  GLUT.DFb<-GLUT.DFb+Tnoise #count matrix with noise only
  GLUT.DFbb<-GLUT.DFb+GLUT.DFb.bias #count matrix with noise and change in expression
  GLUT.RN<-GLUT.counts
  GLUT.BN<-GLUT.counts
  GLUT.RN[modexGenes,]<-GLUT.DFb[modexGenes,] #full count matrix with noise only
  GLUT.BN[modexGenes,]<-GLUT.DFbb[modexGenes,] #full count matrix with noise and change in expression
  #Creating and merging 6 Seurat objects with the following identities:
  #"GLUT.modex" original GLUT
  #"GLUT.DFb" original+noise
  #"GLUT.DFbb" original+noise+change in expression
  #"GLUT.SCT" original+SCT
  #"GLUT.RN.SCT" original+noise+SCT
  #"GLUT.BN.SCT" original+noise+change-in-expression+SCT
  GLUT.nonoise<-CreateSeuratObject(counts = GLUT.counts, min.cells = 0, min.features = 0)
  GLUT.rnoise<-CreateSeuratObject(counts = GLUT.RN, min.cells = 0, min.features = 0)
  GLUT.bnoise<-CreateSeuratObject(counts = GLUT.BN, min.cells = 0, min.features = 0)
  GLUT.nonoise[["percent.mt"]] <- PercentageFeatureSet(GLUT.nonoise, pattern = "^mt-")
  GLUT.rnoise[["percent.mt"]] <- PercentageFeatureSet(GLUT.rnoise, pattern = "^mt-")
  GLUT.bnoise[["percent.mt"]] <- PercentageFeatureSet(GLUT.bnoise, pattern = "^mt-")
  GLUT.SCT<-SCTransform(GLUT.nonoise,vars.to.regress = "percent.mt", verbose = FALSE)
  GLUT.RN.SCT<-SCTransform(GLUT.rnoise,vars.to.regress = "percent.mt", verbose = FALSE)
  GLUT.BN.SCT<-SCTransform(GLUT.bnoise,vars.to.regress = "percent.mt", verbose = FALSE)
  GLUT.SCT.mat<-as.matrix(GLUT.SCT[["SCT"]]$counts)
  GLUT.RN.SCT.mat<-as.matrix(GLUT.RN.SCT[["SCT"]]$counts)
  GLUT.BN.SCT.mat<-as.matrix(GLUT.BN.SCT[["SCT"]]$counts)
  GLUT.SCT<-CreateSeuratObject(counts = GLUT.SCT.mat, min.cells = 0, min.features = 0)
  GLUT.RN.SCT<-CreateSeuratObject(counts = GLUT.RN.SCT.mat, min.cells = 0, min.features = 0)
  GLUT.BN.SCT<-CreateSeuratObject(counts = GLUT.BN.SCT.mat, min.cells = 0, min.features = 0)
  Idents(GLUT.nonoise)<-"GLUT.modex"
  Idents(GLUT.rnoise)<-"GLUT.DFb"
  Idents(GLUT.bnoise)<-"GLUT.DFbb"
  Idents(GLUT.SCT)<-"GLUT.SCT"
  Idents(GLUT.RN.SCT)<-"GLUT.RN.SCT"
  Idents(GLUT.BN.SCT)<-"GLUT.BN.SCT"
  GLUT.sens<-JoinLayers(merge(GLUT.nonoise, y=c(GLUT.rnoise,GLUT.bnoise,GLUT.SCT,GLUT.RN.SCT,GLUT.BN.SCT), add.cell.ids = c("C1","C2","C3","C4","C5","C6"), merge.data = "FALSE"))
  GLUT.sensLN<-NormalizeData(GLUT.sens)
  print(paste("Simulation:",i))
  #Performing various pairwise tests
  GLUT.RN.FM <- as.data.frame(FindMarkers(GLUT.sensLN, features=modexGenes, ident.1 = "GLUT.modex",ident.2 = "GLUT.DFb",logfc.threshold = 0,min.pct=0))
  GLUT.BN.FM <- as.data.frame(FindMarkers(GLUT.sensLN, features=modexGenes.mod, ident.1 = "GLUT.modex",ident.2 = "GLUT.DFbb",logfc.threshold = 0,min.pct=0))
  GLUT.RN.DGE <- DGE.2samples(GLUT.sens, features = modexGenes, ident.1 = "GLUT.modex",ident.2 = "GLUT.DFb", min.count=0, max.pval = 1)
  GLUT.BN.DGE <- DGE.2samples(GLUT.sens, features = modexGenes.mod, ident.1 = "GLUT.modex",ident.2 = "GLUT.DFbb", min.count=0, max.pval = 1)
  GLUT.RN.SCT.DGE <- DGE.2samples(GLUT.sens, features = modexGenes, ident.1 = "GLUT.SCT",ident.2 = "GLUT.RN.SCT", min.count=0, max.pval = 1)
  GLUT.BN.SCT.DGE <- DGE.2samples(GLUT.sens, features = modexGenes.mod, ident.1 = "GLUT.SCT",ident.2 = "GLUT.BN.SCT", min.count=0, max.pval = 1)
  GLUT.RN.T <- IterWghtTtest(GLUT.sens, features = modexGenes, ident.1 = "GLUT.modex",ident.2 = "GLUT.DFb", min.count=0, max.pval = 1, icc=1)
  GLUT.BN.T <- IterWghtTtest(GLUT.sens, features = modexGenes.mod, ident.1 = "GLUT.modex",ident.2 = "GLUT.DFbb", min.count=0, max.pval = 1, icc=1)
  GLUT.RN.FMT <- as.data.frame(FindMarkers(GLUT.sensLN, features=modexGenes, test.use="t",ident.1 = "GLUT.modex",ident.2 = "GLUT.DFb",logfc.threshold = 0,min.pct=0))
  GLUT.BN.FMT <- as.data.frame(FindMarkers(GLUT.sensLN, features=modexGenes.mod, test.use="t",ident.1 = "GLUT.modex",ident.2 = "GLUT.DFbb",logfc.threshold = 0,min.pct=0))
  #Results of pairwise tests: 05s indicates p<0.05 but incorrect sign of log2FC, true indicates matching p-value and |log2FC|>=0.1
  results.RN$FM05[i]<-length(GLUT.RN.FM[GLUT.RN.FM[,1]<0.05,1])
  results.BN$FM05s[i]<-length(GLUT.BN.FM[(GLUT.BN.FM[,2]>0)&(GLUT.BN.FM[,1]<0.05),1])
  results.RN$FM001[i]<-length(GLUT.RN.FM[GLUT.RN.FM[,1]<0.001,1])
  results.BN$FM001[i]<-length(GLUT.BN.FM[GLUT.BN.FM[,1]<0.001,1])
  results.BN$FM05.true[i]<-length(GLUT.BN.FM[(GLUT.BN.FM[,2]<=-0.1)&(GLUT.BN.FM[,1]<0.05),1])
  results.BN$FM001.true[i]<-length(GLUT.BN.FM[(GLUT.BN.FM[,2]<=-0.1)&(GLUT.BN.FM[,1]<0.001),1])
  results.RN$T05[i]<-length(GLUT.RN.T[GLUT.RN.T[,2]<0.05,1])
  results.BN$T05s[i]<-length(GLUT.BN.T[(GLUT.BN.T[,2]<0.05)&(GLUT.BN.T[,1]>0),1])
  results.RN$T001[i]<-length(GLUT.RN.T[GLUT.RN.T[,2]<0.001,1])
  results.BN$T001[i]<-length(GLUT.BN.T[GLUT.BN.T[,2]<0.001,1])
  results.BN$T05.true[i]<-length(GLUT.BN.T[(GLUT.BN.T[,2]<0.05)&(GLUT.BN.T[,1]<=-0.1),1])
  results.BN$T001.true[i]<-length(GLUT.BN.T[(GLUT.BN.T[,2]<0.001)&(GLUT.BN.T[,1]<=-0.1),1])
  results.RN$FMT05[i]<-length(GLUT.RN.FMT[GLUT.RN.FMT[,1]<0.05,1])
  results.BN$FMT05s[i]<-length(GLUT.BN.FMT[(GLUT.BN.FMT[,2]>0)&(GLUT.BN.FMT[,1]<0.05),1])
  results.RN$FMT001[i]<-length(GLUT.RN.FMT[GLUT.RN.FMT[,1]<0.001,1])
  results.BN$FMT001[i]<-length(GLUT.BN.FMT[GLUT.BN.FMT[,1]<0.001,1])
  results.BN$FMT05.true[i]<-length(GLUT.BN.FMT[(GLUT.BN.FMT[,2]<=-0.1)&(GLUT.BN.FMT[,1]<0.05),1])
  results.BN$FMT001.true[i]<-length(GLUT.BN.FMT[(GLUT.BN.FMT[,2]<=-0.1)&(GLUT.BN.FMT[,1]<0.001),1])
  results.RN$CHIIWT05[i]<-length(GLUT.RN.DGE[((GLUT.RN.DGE$p.value<0.05)&(GLUT.RN.DGE$Chi2.p.value<0.05)),1])
  results.BN$CHIIWT05s[i]<-length(GLUT.BN.DGE[((GLUT.BN.DGE$p.value<0.05)&(GLUT.BN.DGE$Chi2.p.value<0.05))&(GLUT.BN.DGE$log2FC>0),1])
  results.RN$CHIIWT001[i]<-length(GLUT.RN.DGE[((GLUT.RN.DGE$p.value<0.001)&(GLUT.RN.DGE$Chi2.p.value<0.05)),1])
  results.BN$CHIIWT001[i]<-length(GLUT.BN.DGE[((GLUT.BN.DGE$p.value<0.001)&(GLUT.BN.DGE$Chi2.p.value<0.05)),1])
  results.RN$SCT05[i]<-length(GLUT.RN.SCT.DGE[((GLUT.RN.SCT.DGE$p.value<0.05)&(GLUT.RN.SCT.DGE$Chi2.p.value<0.05)),1])
  results.BN$SCT05s[i]<-length(GLUT.BN.SCT.DGE[(((GLUT.BN.SCT.DGE$p.value<0.05)&(GLUT.BN.SCT.DGE$Chi2.p.value<0.05))&(GLUT.BN.SCT.DGE$log2FC>0)),1])
  results.RN$SCT001[i]<-length(GLUT.RN.SCT.DGE[((GLUT.RN.SCT.DGE$p.value<0.001)&(GLUT.RN.SCT.DGE$Chi2.p.value<0.05)),1])
  results.BN$SCT001[i]<-length(GLUT.BN.SCT.DGE[((GLUT.BN.SCT.DGE$p.value<0.001)&(GLUT.BN.SCT.DGE$Chi2.p.value<0.05)),1])
  results.BN$CHIIWT05.true[i]<-length(GLUT.BN.DGE[((GLUT.BN.DGE$p.value<0.05)&(GLUT.BN.DGE$Chi2.p.value<0.05))&(GLUT.BN.DGE$log2FC<=-0.1),1])
  results.BN$CHIIWT001.true[i]<-length(GLUT.BN.DGE[((GLUT.BN.DGE$p.value<0.001)&(GLUT.BN.DGE$Chi2.p.value<0.05))&(GLUT.BN.DGE$log2FC<=-0.1),1])
  results.BN$SCT05.true[i]<-length(GLUT.BN.SCT.DGE[((GLUT.BN.SCT.DGE$p.value<0.05)&(GLUT.BN.SCT.DGE$Chi2.p.value<0.05))&(GLUT.BN.SCT.DGE$log2FC<=-0.1),1])
  results.BN$SCT001.true[i]<-length(GLUT.BN.SCT.DGE[((GLUT.BN.SCT.DGE$p.value<0.001)&(GLUT.BN.SCT.DGE$Chi2.p.value<0.05))&(GLUT.BN.SCT.DGE$log2FC<=-0.1),1])
}
#Saving simulation results as csv files
if(sim=="Fig.3") {
  write.csv(results.RN,"results.RN.DF15.1000_300.30_2sample.csv")
  write.csv(results.BN,"results.BN05.DF15.1000_300.30_2sample.csv")
  write.csv(GLUT.RN.FM,"GLUT.RN.FM.DF15.1000_300.30_2sample.csv")
  write.csv(GLUT.RN.FMT,"GLUT.RN.FMT.DF15.1000_300.30_2sample.csv")
  write.csv(GLUT.RN.DGE,"GLUT.RN.DGE.DF15.1000_300.30_2sample.csv")
  write.csv(GLUT.RN.T,"GLUT.RN.T.DF15.1000_300.30_2sample.csv")
  write.csv(GLUT.BN.FM,"GLUT.BN.FM.DF15.1000_300.30_2sample.csv")
  write.csv(GLUT.BN.FMT,"GLUT.BN.FMT.DF15.1000_300.30_2sample.csv")
  write.csv(GLUT.BN.DGE,"GLUT.BN.DGE.DF15.1000_300.30_2sample.csv")
  write.csv(GLUT.BN.T,"GLUT.BN.T.DF15.1000_300.30_2sample.csv")
  write.csv(GLUT.RN.SCT.DGE,"GLUT.RN.SCT.DGE.DF15.1000_300.30_2sample.csv")
  write.csv(GLUT.BN.SCT.DGE,"GLUT.BN.SCT.DGE.DF15.1000_300.30_2sample.csv")
}
if(sim=="Fig.4") {
  write.csv(results.RN,"results.RN.DF15.1000_300.0.3_3_2sample.csv")
  write.csv(results.BN,"results.BN05.DF15.1000_300.0.3_3_2sample.csv")
  write.csv(GLUT.RN.FM,"GLUT.RN.FM.DF15.1000_300.0.3_3_2sample.csv")
  write.csv(GLUT.RN.FMT,"GLUT.RN.FMT.DF15.1000_300.0.3_3_2sample.csv")
  write.csv(GLUT.RN.DGE,"GLUT.RN.DGE.DF15.1000_300.0.3_3_2sample.csv")
  write.csv(GLUT.RN.T,"GLUT.RN.T.DF15.1000_300.0.3_3_2sample.csv")
  write.csv(GLUT.BN.FM,"GLUT.BN.FM.DF15.1000_300.0.3_3_2sample.csv")
  write.csv(GLUT.BN.FMT,"GLUT.BN.FMT.DF15.1000_300.0.3_3_2sample.csv")
  write.csv(GLUT.BN.DGE,"GLUT.BN.DGE.DF15.1000_300.0.3_3_2sample.csv")
  write.csv(GLUT.BN.T,"GLUT.BN.T.DF15.1000_300.0.3_3_2sample.csv")
  write.csv(GLUT.RN.SCT.DGE,"GLUT.RN.SCT.DGE.DF15.1000_300.0.3_3_2sample.csv")
  write.csv(GLUT.BN.SCT.DGE,"GLUT.BN.SCT.DGE.DF15.1000_300.0.3_3_2sample.csv")
}
#
######################################################
#'* FIGURE 5*
######################################################
#
GLUT.QCN<-NormalizeData(GLUT.QC, normalization.method = "RC")
GLUT.counts<-GLUT.QC[["RNA"]]$counts
GLUT.Ni<-colSums(GLUT.counts)
GLUT.Xi<-as.matrix(GLUT.QCN[["RNA"]]$data)/10000
GLUT.cov<-vector()
#Selecting moderate expression genes
GLUTmodex.counts<-GLUT.counts[rowSums(GLUT.counts)/ncol(GLUT.counts)>0.3,]
GLUTmodex.counts<-GLUTmodex.counts[rowSums(GLUTmodex.counts)/ncol(GLUTmodex.counts)<3,]
#Selecting 500 moderate expression genes with the largest coefficient of variation
for (gene in rownames(GLUTmodex.counts)) {GLUT.cov[gene]=sd(GLUT.Xi[gene,])/mean(GLUT.Xi[gene,])}
GLUT.cov<-sort(GLUT.cov,decreasing = TRUE)
modexGenes1<-names(GLUT.cov[c(1:500)])
#Randomly selecting additional 500 moderate expression genes
modexGenes2<-sample(setdiff(rownames(GLUTmodex.counts),modexGenes1),500)
modexGenes<-c(modexGenes1,modexGenes2)
#Defining dataframe for counting false positive DGE
results.PB.RN<-data.frame(DES05=numeric(10),DES001=numeric(10),T05=numeric(10),T001=numeric(10),
                          WT05=numeric(10),WT001=numeric(10), IWT05=numeric(10),IWT001=numeric(10))
#Defining dataframe counting true positive DGE
results.PB.BN<-data.frame(DES05=numeric(10),DES001=numeric(10),T05=numeric(10),T001=numeric(10),
                          WT05=numeric(10),WT001=numeric(10), IWT05=numeric(10),IWT001=numeric(10))
# Main simulation loop
for (sim in c(1:10)){                  
  print(paste("Simulation:",sim))
  set.seed(sim)
  modexGenes.mod<-sample(modexGenes,300)   #300 genes for modifying expression
  modexGenes.mod1<-sample(modexGenes.mod,150) #Splitting the 300 genes into 2 150 gene subgroups
  modexGenes.mod2<-setdiff(modexGenes.mod,modexGenes.mod1)
  GLUTmod2.counts<-GLUT.counts[modexGenes,]
  modexCells<-colnames(GLUTmodex.counts)
  Ngenes<-length(modexGenes)
  Ncells<-ncol(GLUT.counts)
  GLUT.PB<-list()
  #Creating 5 unaltered and 5 altered datasets
  for (i in c(1:10)){
    print(paste0("Generating sample:",i))
    DF<-sqrt(15)
    if(i>5){
      modCells<-sample(as.vector(modexCells),round(0.2*(i-5)*Ncells))
      GLUT.DFb<-GLUTmod2.counts[,modCells]
    }
    else {
      modCells<-sample(as.vector(modexCells),round(0.2*i*Ncells))
      GLUT.DFb<-GLUTmod2.counts[,modCells]
    }
    GLUT.RBN<-GLUT.counts[,modCells]
    GL.Ni<-as.vector(colSums(GLUT.RBN))
    GL.N<-sum(GL.Ni)
    GL.AC<-rowSums(GLUT.DFb)
    Nc<-ncol(GLUT.DFb)
    Nr<-nrow(GLUT.DFb)
    #Bias matrix
    GLUT.DFb.bias<-matrix(0,ncol= Nc,nrow = Nr)
    rownames(GLUT.DFb.bias)<-modexGenes
    colnames(GLUT.DFb.bias)<-colnames(GLUT.DFb)
    #Adding bias (change in gene expression)
    set.seed(10+i+10*sim)
    if (i>5) {
      for (gene in modexGenes.mod1){
        GLUT.DFb[gene,]<-GLUT.DFb[gene,]+round(0.3*GLUT.DFb[gene,]) #set sign to "-" to have 150 downregulated and 150 upregulated genes
      }
      for (gene in modexGenes.mod2){
        GLUT.DFb[gene,]<-GLUT.DFb[gene,]+round(0.3*GLUT.DFb[gene,])
      }
    }
    Unoise<-matrix(runif(Nc*Nr,min=-1,max=1),ncol=Nc,nrow=Nr)
    Tnoise<-round(Unoise*(DF-(DF-1)/(1+0.1*(GLUT.DFb-1)))*sqrt(GLUT.DFb))
    GLUT.DFb<-GLUT.DFb+Tnoise
    GLUT.RBN[modexGenes,]<-GLUT.DFb[modexGenes,]
    GLUT.PB[[i]]<-CreateSeuratObject(counts = GLUT.RBN, min.cells = 0, min.features = 0)
    Idents(GLUT.PB[[i]])<-paste("GLUT_",i,sep = "")
  }
  #Creating multi-sample dataset
  GLUT.MS<-JoinLayers(merge(GLUT.PB[[1]], y=GLUT.PB[c(2:10)], add.cell.ids=c("C1","C2","C3","C4","C5","C6","C7","C8","C9","C10")))
  #Classical PB-DESeq2 data analysis
  GLUT.MS.AC<-matrix(0,nrow = nrow(GLUT.MS),ncol =length(levels(GLUT.MS@active.ident)))
  rownames(GLUT.MS.AC)<-rownames(GLUT.MS[["RNA"]]$counts)
  samples<-levels(GLUT.MS@active.ident)
  colnames(GLUT.MS.AC)<-samples
  for (i in c(1:ncol(GLUT.MS.AC))) {
    GLUT.MS.Ci<-as.matrix(subset(GLUT.MS, idents = samples[i])[["RNA"]]$counts)
    GLUT.MS.AC[,i]<-rowSums(GLUT.MS.Ci)
  }
  GLUT.5RN<-CreateSeuratObject(counts = GLUT.MS.AC[,c(1:5)], min.cells = 0, min.features = 0)
  Idents(GLUT.5RN)<-"5RN"
  GLUT.5BN<-CreateSeuratObject(counts = GLUT.MS.AC[,c(6:10)], min.cells = 0, min.features = 0)
  Idents(GLUT.5BN)<-"5BN"
  GLUT.5RN.5BN.DES<-JoinLayers(merge(GLUT.5RN, y=GLUT.5BN, merge.data = "FALSE"))
  GLUT.5RN.5BN.DESN<-NormalizeData(GLUT.5RN.5BN.DES, normalization.method = "RC")
  GLUT.5RN.5BN.DESeq2<-as.data.frame(FindMarkers(GLUT.5RN.5BN.DESN, features= modexGenes, ident.1 = "5RN",ident.2 = "5BN",test.use="DESeq2", logfc.threshold = 0,min.pct=0.1))
  GLUT.5RN.5BN.DESeq2.bias<-GLUT.5RN.5BN.DESeq2[modexGenes.mod,]
  modexGenes.notmod<-setdiff(modexGenes,modexGenes.mod)
  GLUT.5RN.5BN.DESeq2.noise<-GLUT.5RN.5BN.DESeq2[modexGenes.notmod,]
  GLUT.5RN.5BN.DESeq2.bias.05<-GLUT.5RN.5BN.DESeq2.bias[(GLUT.5RN.5BN.DESeq2.bias[,1]<0.05)&(abs(GLUT.5RN.5BN.DESeq2.bias[,2])>=0.1),]
  GLUT.5RN.5BN.DESeq2.bias.001<-GLUT.5RN.5BN.DESeq2.bias.05[GLUT.5RN.5BN.DESeq2.bias.05[,1]<0.001,]
  GLUT.5RN.5BN.DESeq2.noise.05<-GLUT.5RN.5BN.DESeq2.noise[GLUT.5RN.5BN.DESeq2.noise[,1]<0.05,]
  GLUT.5RN.5BN.DESeq2.noise.001<-GLUT.5RN.5BN.DESeq2.noise[GLUT.5RN.5BN.DESeq2.noise[,1]<0.001,]
  #Weighted t-test for weighted average counts
  GLUT.5RN.5BN.bias<-DGE.MultiSample(GLUT.MS,samples.1=samples[c(1:5)],samples.2=samples[c(6:10)],
                                     features=modexGenes.mod,t.test=FALSE, min.pct=0.03,fc.thr=1,max.pval=1,icc="i")
  GLUT.5RN.5BN.noise<-DGE.MultiSample(GLUT.MS,samples.1=samples[c(1:5)],samples.2=samples[c(6:10)],
                                      features=modexGenes.notmod,t.test=FALSE, min.pct=0.03,fc.thr=1,max.pval=1,icc="i")
  GLUT.5RN.5BN.bias.05<-GLUT.5RN.5BN.bias$DGE[(GLUT.5RN.5BN.bias$DGE[,2]<0.05)&(abs(GLUT.5RN.5BN.bias$DGE[,1])>=0.1),]
  GLUT.5RN.5BN.bias.001<-GLUT.5RN.5BN.bias.05[GLUT.5RN.5BN.bias.05[,2]<0.001,]
  GLUT.5RN.5BN.noise.05<-GLUT.5RN.5BN.noise$DGE[GLUT.5RN.5BN.noise$DGE[,2]<0.05,]
  GLUT.5RN.5BN.noise.001<-GLUT.5RN.5BN.noise$DGE[GLUT.5RN.5BN.noise$DGE[,2]<0.001,]
  #Weighted t-test for aggregate counts
  GLUT.5RN.5BN.WT.bias<-DGE.MultiSample(GLUT.MS,samples.1=samples[c(1:5)],samples.2=samples[c(6:10)],
                                        features=modexGenes.mod,t.test=FALSE, min.pct=0.03, fc.thr=1,max.pval=1,icc=0)
  GLUT.5RN.5BN.WT.noise<-DGE.MultiSample(GLUT.MS,samples.1=samples[c(1:5)], min.pct=0.03,samples.2=samples[c(6:10)],
                                         features=modexGenes.notmod,t.test=FALSE, fc.thr=1,max.pval=1,icc=0)
  GLUT.5RN.5BN.WT.bias.05<-GLUT.5RN.5BN.WT.bias$DGE[(GLUT.5RN.5BN.WT.bias$DGE[,2]<0.05)&(abs(GLUT.5RN.5BN.WT.bias$DGE[,1])>=0.1),]
  GLUT.5RN.5BN.WT.bias.001<-GLUT.5RN.5BN.WT.bias.05[GLUT.5RN.5BN.WT.bias.05[,2]<0.001,]
  GLUT.5RN.5BN.WT.noise.05<-GLUT.5RN.5BN.WT.noise$DGE[GLUT.5RN.5BN.WT.noise$DGE[,2]<0.05,]
  GLUT.5RN.5BN.WT.noise.001<-GLUT.5RN.5BN.WT.noise$DGE[GLUT.5RN.5BN.WT.noise$DGE[,2]<0.001,]
  #Common t-test for aggregate counts
  GLUT.5RN.5BN.T.bias<-DGE.MultiSample(GLUT.MS,samples.1=samples[c(1:5)],samples.2=samples[c(6:10)],
                                       features=modexGenes.mod,t.test=TRUE, min.pct=0.03, fc.thr=1,max.pval=1,icc=0)
  GLUT.5RN.5BN.T.noise<-DGE.MultiSample(GLUT.MS,samples.1=samples[c(1:5)],samples.2=samples[c(6:10)],
                                        features=modexGenes.notmod,t.test=TRUE, min.pct=0.03, fc.thr=1,max.pval=1,icc=0)
  GLUT.5RN.5BN.T.bias.05<-GLUT.5RN.5BN.T.bias$DGE[(GLUT.5RN.5BN.T.bias$DGE[,2]<0.05)&(abs(GLUT.5RN.5BN.T.bias$DGE[,1])>=0.1),]
  GLUT.5RN.5BN.T.bias.001<-GLUT.5RN.5BN.T.bias.05[GLUT.5RN.5BN.T.bias.05[,2]<0.001,]
  GLUT.5RN.5BN.T.noise.05<-GLUT.5RN.5BN.T.noise$DGE[GLUT.5RN.5BN.T.noise$DGE[,2]<0.05,]
  GLUT.5RN.5BN.T.noise.001<-GLUT.5RN.5BN.T.noise$DGE[GLUT.5RN.5BN.T.noise$DGE[,2]<0.001,]
  #Assembling dataframes for differentially expressed genes
  results.PB.RN$DES05[sim]<-nrow(GLUT.5RN.5BN.DESeq2.noise.05)
  results.PB.RN$DES001[sim]<-nrow(GLUT.5RN.5BN.DESeq2.noise.001)
  results.PB.BN$DES05[sim]<-nrow(GLUT.5RN.5BN.DESeq2.bias.05)
  results.PB.BN$DES001[sim]<-nrow(GLUT.5RN.5BN.DESeq2.bias.001)
  results.PB.RN$T05[sim]<-nrow(GLUT.5RN.5BN.T.noise.05)
  results.PB.RN$T001[sim]<-nrow(GLUT.5RN.5BN.T.noise.001)
  results.PB.BN$T05[sim]<-nrow(GLUT.5RN.5BN.T.bias.05)
  results.PB.BN$T001[sim]<-nrow(GLUT.5RN.5BN.T.bias.001)
  results.PB.RN$WT05[sim]<-nrow(GLUT.5RN.5BN.WT.noise.05)
  results.PB.RN$WT001[sim]<-nrow(GLUT.5RN.5BN.WT.noise.001)
  results.PB.BN$WT05[sim]<-nrow(GLUT.5RN.5BN.WT.bias.05)
  results.PB.BN$WT001[sim]<-nrow(GLUT.5RN.5BN.WT.bias.001)
  results.PB.RN$IWT05[sim]<-nrow(GLUT.5RN.5BN.noise.05)
  results.PB.RN$IWT001[sim]<-nrow(GLUT.5RN.5BN.noise.001)
  results.PB.BN$IWT05[sim]<-nrow(GLUT.5RN.5BN.bias.05)
  results.PB.BN$IWT001[sim]<-nrow(GLUT.5RN.5BN.bias.001)
}
#Saving simulation results as csv files
write.csv(results.PB.RN,"results.PB.RN.DF15.1000_300.03_3_30%.csv")
write.csv(results.PB.BN,"results.PB.BN.DF15.1000_300.03_3_30%.csv")
write.csv(GLUT.5RN.5BN.bias,"GLUT.5RN.5BN.bias.DF15.1000_300.03_3_30%.csv")
write.csv(GLUT.5RN.5BN.bias.05,"GLUT.5RN.5BN.bias.05.DF15.1000_300.03_3_30%.csv")
write.csv(GLUT.5RN.5BN.bias.001,"GLUT.5RN.5BN.bias.001.DF15.1000_300.03_3_30%.csv")
write.csv(GLUT.5RN.5BN.noise.05,"GLUT.5RN.5BN.noise.05.DF15.1000_300.03_3_30%_2.csv")
write.csv(GLUT.5RN.5BN.noise.001,"GLUT.5RN.5BN.noise.001.DF15.1000_300.03_3_30%_2.csv")
write.csv(GLUT.5RN.5BN.WT.bias,"GLUT.5RN.5BN.WT.bias.DF15.1000_300.03_3_30%_2.csv")
write.csv(GLUT.5RN.5BN.WT.bias.05,"GLUT.5RN.5BN.WT.bias.05.DF15.1000_300.03_3_30%_2.csv")
write.csv(GLUT.5RN.5BN.WT.bias.001,"GLUT.5RN.5BN.WT.bias.001.DF15.1000_300.03_3_30%_2.csv")
write.csv(GLUT.5RN.5BN.WT.noise.05,"GLUT.5RN.5BN.WT.noise.05.DF15.1000_300.03_3_30%_2.csv")
write.csv(GLUT.5RN.5BN.WT.noise.001,"GLUT.5RN.5BN.WT.noise.001.DF15.1000_300.03_3_30%_2.csv")
write.csv(GLUT.5RN.5BN.T.bias,"GLUT.5RN.5BN.T.bias.DF15.1000_300.03_3_30%_2.csv")
write.csv(GLUT.5RN.5BN.T.bias.05,"GLUT.5RN.5BN.T.bias.05.DF15.1000_300.03_3_30%_2.csv")
write.csv(GLUT.5RN.5BN.T.bias.001,"GLUT.5RN.5BN.T.bias.001.DF15.1000_300.03_3_30%_2.csv")
write.csv(GLUT.5RN.5BN.T.noise.05,"GLUT.5RN.5BN.T.noise.05.DF15.1000_300.03_3_30%_2.csv")
write.csv(GLUT.5RN.5BN.T.noise.001,"GLUT.5RN.5BN.T.noise.001.DF15.1000_300.03_3_30%_2.csv")
write.csv(GLUT.5RN.5BN.DESeq2.bias,"GLUT.5RN.5BN.DESeq2.bias.DF15.1000_300.03_3_30%_2.csv")
write.csv(GLUT.5RN.5BN.DESeq2.bias.05,"GLUT.5RN.5BN.DESeq2.bias.05.DF15.1000_300.03_3_30%_2.csv")
write.csv(GLUT.5RN.5BN.DESeq2.bias.001,"GLUT.5RN.5BN.DESeq2.bias.001.DF15.1000_300.03_3_30%_2.csv")
write.csv(GLUT.5RN.5BN.DESeq2.noise.05,"GLUT.5RN.5BN.DESeq2.noise.05.DF15.1000_300.03_3_30%_2.csv")
write.csv(GLUT.5RN.5BN.DESeq2.noise.001,"GLUT.5RN.5BN.DESeq2.noise.001.DF15.1000_300.03_3_30%_2.csv")
#
######################################################
#'* FIGURE 6*
######################################################
# Read assembled spRNASeq (VisiumHD) data for 3 proliferating chondrocyte regions in E18 tibia
E18tib.PCN<-readRDS("E18tib.PCN.rds")
# function for visiualizing the data
GeneMap <- function(object,map.features="nCount_Spatial.008um",map.limits=c(0,10000),pt.size=1,map.breaks=c(0,2500,5000,7500,10000),
                    map.colors= c("#00008b","#00ff8b","yellow","#cf2222"),image.crop=FALSE,colorbar.position="bottom",legend.size=1){
  SpatialFeaturePlot(object, features = map.features, pt.size.factor = pt.size,crop = image.crop)+
    ggplot2::scale_fill_gradientn(limits=map.limits,breaks=map.breaks,colors=map.colors)+
    theme(legend.position = colorbar.position, legend.key.width = unit(1*legend.size,"cm"),
          legend.key.height = unit(0.6*legend.size,"cm"),
          legend.title = element_text(size=14*legend.size),
          legend.text = element_text(size=12*legend.size))
}
# Visualization of the data
GM1 <-GeneMap(subset(E18tib.PCN, idents="Top"),map.features = "Col2a1",pt.size = 1,map.limits = c(0,2000),map.breaks = c(0,500,1000,1500))
GM2 <-GeneMap(subset(E18tib.PCN, idents="Mid"),map.features = "Col2a1",pt.size = 1,map.limits = c(0,2000),map.breaks = c(0,500,1000,1500))
GM3 <-GeneMap(subset(E18tib.PCN, idents="Bot"),map.features = "Col2a1",pt.size = 1,map.limits = c(0,2000),map.breaks = c(0,500,1000,1500))
GM1|GM2|GM3
# Renaming active assay from "Spatial.008um" to "RNA"
E18tib.PCN=RenameAssays(object = E18tib.PCN, Spatial.008um="RNA")
# Genes detected by weighted DGE analysis
E18top.bot<-DGE.2samples(E18tib.PCN, ident.1 = "Top",ident.2 = "Bot", min.pct=0.1, max.pval = 0.05, icc = 0)
E18top.mid<-DGE.2samples(E18tib.PCN, ident.1 = "Top",ident.2 = "Mid", min.pct=0.1, max.pval = 0.05, icc = 0)
E18bot.mid<-DGE.2samples(E18tib.PCN, ident.1 = "Bot",ident.2 = "Mid", min.pct=0.1, max.pval = 0.05, icc = 0)
genes<-intersect(rownames(E18top.bot),rownames(E18top.mid))
genes<-intersect(genes,rownames(E18bot.mid))
# Genes detected by default Seurat analysis (Wilcoxon test)
E18top.bot.S=as.data.frame(FindMarkers(E18tib.PCN, ident.1 = "Top",ident.2 = "Bot", min.pct=0.1, logfc.threshold = 0))
E18top.bot.S=E18top.bot.S[E18top.bot.S$p_val<0.05,]
E18top.mid.S=as.data.frame(FindMarkers(E18tib.PCN, ident.1 = "Top",ident.2 = "Mid", min.pct=0.1, logfc.threshold = 0))
E18top.mid.S=E18top.mid.S[E18top.mid.S$p_val<0.05,]
E18bot.mid.S=as.data.frame(FindMarkers(E18tib.PCN, ident.1 = "Bot",ident.2 = "Mid", min.pct=0.1, logfc.threshold = 0))
E18bot.mid.S=E18bot.mid.S[E18bot.mid.S$p_val<0.05,]
genes.S<-intersect(rownames(E18top.bot.S),rownames(E18top.mid.S))
genes.S<-intersect(genes.S,rownames(E18bot.mid.S))

